lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -374,75 +374,94 @@ static bool hasSideEffect(Operation &op) {
 // TODO: Move to the op.
 static unsigned distributionRatio = 32;
 
-void mlir::vector_ext::distributeTransferWrite(
-    OpBuilder &builder, WarpSingleLaneOp op,
-    std::function<AffineMap(vector::TransferWriteOp)> distributionMapFn) {
-  vector::TransferWriteOp writeOp;
-  while (1) {
-    // Find the first transfer_write from the end of the block.
-    for (Operation &elementOp : llvm::reverse(op.getBody()->getOperations())) {
-      writeOp = dyn_cast<vector::TransferWriteOp>(elementOp);
-      if (writeOp)
-        break;
-      if (hasSideEffect(elementOp))
-        return;
-    }
-    if (!writeOp)
-      return;
+struct WarpOpTransferWrite : public OpRewritePattern<vector::TransferWriteOp> {
+  WarpOpTransferWrite(MLIRContext *ctx, DistributionMapFn fn,
+                      PatternBenefit b = 1)
+      : OpRewritePattern<vector::TransferWriteOp>(ctx, b),
+        distributionMapFn(fn) {}
+
+  LogicalResult matchAndRewrite(vector::TransferWriteOp writeOp,
+                                PatternRewriter &rewriter) const override {
+    auto warpOp = dyn_cast<WarpSingleLaneOp>(writeOp->getParentOp());
+    if (!warpOp)
+      return failure();
+
+    // There must be no op with a side effect after writeOp.
+    // TODO: What about ops with regions?
```

Comment:
`hasSideEffect` checks for recursive side effects so I don't think it should be problem